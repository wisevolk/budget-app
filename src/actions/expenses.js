// ADD_EXPENSE
import { v4 as uuid } from "uuid";

export const addExpense = ({
                        description = "",
                        note = "",
                        amount = 0,
                        createdAt = 0
                    } = {}) => ({
    type: "ADD_EXPENSE",
    expense: {
        id: uuid(),
        description,
        note,
        amount,
        createdAt
    }
});
// REMOVE_EXPENSE
export const removeExpense = ({ id = uuid.NIL } = {}) => ({
    type: "REMOVE_EXPENSE",
    id
});

// EDIT_EXPENSE
export const editExpense = (id = uuid.NIL, updates) => ({
    type: "EDIT_EXPENSE",
    id,
    updates
});
