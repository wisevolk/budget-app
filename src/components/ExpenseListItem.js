import React from "react";
import { Link } from "react-router-dom";

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
    <div>
        <Link to={ `/edit/${ id }` }><p>{ description }</p></Link>
        <p>{ amount } - { createdAt }</p>
        {/*<button onClick={ (e) => {*/ }
        {/*    dispatch(removeExpense({ id }))*/ }
        {/*} }>remove*/ }
        {/*</button>*/ }
    </div>
);

export default ExpenseListItem;
