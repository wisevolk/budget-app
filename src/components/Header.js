import React from "react";
import { NavLink } from "react-router-dom";

const Header = () => (
    <div>
        <h3>Budget App!</h3>
        <NavLink to={ "/" } activeClassName={ "is-active" } exact>Home</NavLink>
        <NavLink to={ "/create" } activeClassName={ "is-active" }>Create new expense</NavLink>
        <NavLink to={ "/help" } activeClassName={ "is-active" }>Help</NavLink>
    </div>
);

export default Header;
