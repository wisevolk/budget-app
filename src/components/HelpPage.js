import React from "react";

const HelpPage = () => (
    <div>
        <p>From HelpPage component</p>
    </div>
);

export default HelpPage;
