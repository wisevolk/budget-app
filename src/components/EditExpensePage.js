import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { editExpense, removeExpense } from "../actions/expenses";

export class EditExpensePage extends React.Component {
    // console.log(props)
    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push("/");
    }

    onRemove = (e) => {
        this.props.removeExpense({ id: this.props.expense.id });
        this.props.history.push("/");
    }

    render() {
        return (
            <div>
                <ExpenseForm
                    expense={ this.props.expense }
                    onSubmit={ this.onSubmit }
                />
                <button onClick={ this.onRemove }>Remove
                </button>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch, props) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: (id) => dispatch(removeExpense( id ))
});

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    // return {
    //     expense: state.expenses.find(expense => {
    //         return expense.id === props.match.params.id;
    //     })
    // };
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
