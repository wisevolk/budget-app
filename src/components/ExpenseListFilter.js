import React from "react";
import { connect } from "react-redux";
import { setEndDate, setStartDate, setTextFilter, sortByAmount, sortByDate } from "../actions/filters";
import { DateRangePicker }  from "react-dates";
import { v4 as uuid } from "uuid";

class ExpenseListFilter extends React.Component {
    state = {
        calendarFocused: null
    }

    render() {
        return (
            <div>
                <input
                    type="text"
                    value={ this.props.filters.text }
                    onChange={ (e => {
                        this.props.dispatch(setTextFilter(e.target.value))
                    }) }
                />
                <select value={ this.props.filters.sortBy } onChange={ (e) => {
                    const sortBy = e.target.value;
                    if( sortBy === "date" ) {
                        this.props.dispatch(sortByDate());
                    } else {
                        this.props.dispatch(sortByAmount());
                    }
                } }>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={ this.props.filters.startDate } // momentPropTypes.momentObj or null,
                    startDateId={ uuid() } // PropTypes.string.isRequired,
                    endDate={ this.props.filters.endDate } // momentPropTypes.momentObj or null,
                    endDateId={ uuid() } // PropTypes.string.isRequired,
                    onDatesChange={ ({ startDate, endDate }) => {
                        this.props.dispatch(setStartDate(startDate));
                        this.props.dispatch(setEndDate(endDate));
                    } } // PropTypes.func.isRequired,
                    focusedInput={ this.state.calendarFocused } // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                    onFocusChange={ calendarFocused => this.setState({ calendarFocused }) } // PropTypes.func.isRequired,
                    showClearDates={ true }
                    numberOfMonths={ 1 }
                    isOutsideRange={ () => false }
                />

            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
};

export default connect(mapStateToProps)(ExpenseListFilter);
