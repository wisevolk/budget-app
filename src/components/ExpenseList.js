import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpense from "../selectors/expenses"

export const ExpenseList = (props) => (
    <div>
        { props.expenses && props.expenses.length !==  0 ? (
            props.expenses.map((expense) => (
                <ExpenseListItem
                    key={ expense.id }
                    { ...expense }
                />
                // console.log(expense);
            ))
        ) : (
            <p>No expenses yet</p>
        ) }
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpense(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseList);
