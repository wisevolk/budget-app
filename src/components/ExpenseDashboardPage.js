import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilter from "./ExpenseListFilter";

const ExpenseDashboardPage = () => (
    <div>
        <p>From Dashboard Component</p>
        <ExpenseListFilter />
        <ExpenseList/>
    </div>
);

export default ExpenseDashboardPage;
