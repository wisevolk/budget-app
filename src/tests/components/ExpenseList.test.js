import React from "react";
import { shallow } from "enzyme";
import { ExpenseList } from "../../components/ExpenseList";
import expenses from "../fixtures/expenses";

test("should generate ExpenseList", () => {
    const wrapper = shallow(<ExpenseList expenses={expenses}/>);
    expect(wrapper).toMatchSnapshot();
});

test("should generate ExpenseList without Expenses", ()=>{
   const wrapper = shallow(<ExpenseList/>);
   expect(wrapper).toMatchSnapshot();
});
