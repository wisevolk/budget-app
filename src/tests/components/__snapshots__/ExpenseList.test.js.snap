// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should generate ExpenseList 1`] = `
<div>
  <ExpenseListItem
    amount={195}
    createdAt={0}
    description="Gum"
    id={0}
    key="0"
    note=""
  />
  <ExpenseListItem
    amount={109500}
    createdAt={-345600000}
    description="Rent"
    id={2}
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={4500}
    createdAt={345600000}
    description="Credit Card"
    id={1}
    key="1"
    note=""
  />
</div>
`;

exports[`should generate ExpenseList without Expenses 1`] = `
<div>
  <p>
    No expenses yet
  </p>
</div>
`;
